File: 18-1.rb
#!/usr/bin/env ruby

require 'numo/narray'

class Map
  def initialize(max_x, max_y, max_z)
    @size_x = max_x + 1
    @size_y = max_y + 1
    @size_z = max_z + 1
    @map = Numo::UInt8.zeros(@size_x, @size_y, @size_z)
  end

  def [](x, y, z)
    @map[x, y, z]
  end

  def []=(x, y, z, val)
    @map[x, y, z] = val
  end

  def count_exposed_faces
    count = 0
    @size_x.times do |x|
      @size_y.times do |y|
        @size_z.times do |z|
          next unless @map[x, y, z] == 1

          count += 1 if x == 0 or @map[x - 1, y, z] == 0
          count += 1 if x == @size_x - 1 or @map[x + 1, y, z] == 0
          count += 1 if y == 0 or @map[x, y - 1, z] == 0
          count += 1 if y == @size_y - 1 or @map[x, y + 1, z] == 0
          count += 1 if z == 0 or @map[x, y, z - 1] == 0
          count += 1 if z == @size_z - 1 or @map[x, y, z + 1] == 0
        end
      end
    end
    count
  end
end

input = File.read('18.input').lines.map(&:strip).map { |l| l.split(',').map(&:to_i) }

max_x = input.map(&:first).max
max_y = input.map { |c| c[1] }.max
max_z = input.map(&:last).max

map = Map.new(max_x, max_y, max_z)

input.each do |coords|
  map[coords[0], coords[1], coords[2]] = 1
end

print map.count_exposed_faces, "\n"

File: 18-1.py
#!/usr/bin/env python

import numpy


class Map:
    def __init__(self, max_x, max_y, max_z):
        self.size_x = max_x + 1
        self.size_y = max_y + 1
        self.size_z = max_z + 1
        self.map = numpy.zeros((self.size_x, self.size_y, self.size_z), dtype=numpy.uint8)

    def count_exposed_faces(self):
        count = 0
        for x in range(self.size_x):
            for y in range(self.size_y):
                for z in range(self.size_z):
                    if self.map[x, y, z] == 1:
                        if x == 0 or self.map[x - 1, y, z] == 0:
                            count += 1
                        if x == self.size_x - 1 or self.map[x + 1, y, z] == 0:
                            count += 1
                        if y == 0 or self.map[x, y - 1, z] == 0:
                            count += 1
                        if y == self.size_y - 1 or self.map[x, y + 1, z] == 0:
                            count += 1
                        if z == 0 or self.map[x, y, z - 1] == 0:
                            count += 1
                        if z == self.size_z - 1 or self.map[x, y, z + 1] == 0:
                            count += 1
        return count


lines = [list(map(lambda c: int(c), line.strip().split(","))) for line in open('18.input').readlines()]

max_x = max([x for x, _, _ in lines])
max_y = max([y for _, y, _ in lines])
max_z = max([z for _, _, z in lines])

obj = Map(max_x, max_y, max_z)

for coords in lines:
    obj.map[coords[0], coords[1], coords[2]] = 1

print(obj.count_exposed_faces())
